src/mpdec.jl

```diff
@@ -0,0 +1,766 @@
+                     0 (sqrt(6sqrt(7oneType) + 21) / 540 - sqrt(7oneType) / 756 - sqrt(42sqrt(7oneType) + 147) / 3780-oneType / 540) (sqrt(7oneType) / 756 - sqrt(21 - 6sqrt(7oneType)) / 540 - sqrt(147 - 42sqrt(7oneType)) / 3780-oneType / 540) (sqrt(7oneType) / 756 + sqrt(21 - 6sqrt(7oneType)) / 540 + sqrt(147 - 42sqrt(7oneType)) / 3780-oneType / 540) (sqrt(42sqrt(7oneType) + 147) / 3780 - sqrt(6sqrt(7oneType) + 21) / 540 - sqrt(7oneType) / 756-oneType / 540) oneType/30]
+        else
+            error("MPDeC requires 2 ≤ K ≤ 10.")
+        end
+    end
+    return nodes, theta
+end
+
+struct MPDeCConstantCache{NType, T, T2} <: OrdinaryDiffEqConstantCache
+    K::Int
+    M::Int
+    nodes::NType
+    theta::T2
+    small_constant::T
+end
+
+# Out-of-place
+function alg_cache(alg::MPDeC, u, rate_prototype, ::Type{uEltypeNoUnits},
+                   ::Type{uBottomEltypeNoUnits}, ::Type{tTypeNoUnits},
+                   uprev, uprev2, f, t, dt, reltol, p, calck,
+                   ::Val{false}) where {uEltypeNoUnits, uBottomEltypeNoUnits, tTypeNoUnits}
+    if !(f isa PDSFunction || f isa ConservativePDSFunction)
+        throw(ArgumentError("MPDeC can only be applied to production-destruction systems"))
+    end
+
+    nodes, theta = get_constant_parameters(alg, uEltypeNoUnits)
+    MPDeCConstantCache(alg.K, alg.M, nodes, theta,
+                       alg.small_constant_function(uEltypeNoUnits))
+end
+
+function initialize!(integrator, cache::MPDeCConstantCache)
+end
+
+# out-of-place
+function build_mpdec_matrix_and_rhs_oop(uprev, m, f, C, p, t, dt, nodes, theta,
+                                        small_constant)
+    N = length(uprev)
+    if f isa PDSFunction
+        # Additional destruction terms 
+        Mmat, rhs = _build_mpdec_matrix_and_rhs_oop(uprev, m, f.p, C, p, t, dt, nodes,
+                                                    theta,
+                                                    small_constant, f.d)
+    else
+        # No additional destruction terms 
+        Mmat, rhs = _build_mpdec_matrix_and_rhs_oop(uprev, m, f.p, C, p, t, dt, nodes,
+                                                    theta,
+                                                    small_constant)
+    end
+
+    if uprev isa StaticArray
+        return SMatrix{N, N}(Mmat), SVector{N}(rhs)
+    else
+        return Mmat, rhs
+    end
+end
+
+# out-of-place for dense arrays
+@muladd function _build_mpdec_matrix_and_rhs_oop(uprev, m, prod, C, p, t, dt, nodes, theta,
+                                                 small_constant, dest = nothing)
+    N, M = size(C)
+    M = M - 1
```

Comment:
It feels a bit odd to me to first decrease `M` by one and after that only use `M + 1` once in this function. Would it be less confusing to delete this line and use `M` instead of `M + 1` below or is there some reason the current form is preferable?